Rewrite of rendering stack

grokys opened the pull request:
Previously we redrew everything on a window each time the window contents changed. This PR implements a new `DeferredRenderer` class (the previous renderer is still available as `ImmediateRenderer`) which renders to a low-level scenegraph on the UI thread which is then renderered to the window on a separate render thread.

This gives us:

- The ability to only draw the controls that have changed
- Proper hit-testing of geometry instead of control bounds
- A dirty-rects debug overlay (currently enabled by default in ControlCatalog and RenderTests - this will be disabled when the PR is ready)

Things left to do:

- [x] `VisualBrush`es are currently broken (don't open the DropDown page in ControlCatalog!)
- [x] Fix DPI scaling on skia
- [x] Fix cairo
- [x] Fix opacity masks
- [x] Fix geometry clips
- [x] Embedding native widgets

Known problems that I'm going to punt to a later PR to prevent this one growing too huge:

- Performance on D2D with controls with transparency is bad (such as disabled buttons). This is because we currently use `RenderTargetBitmap`s for render layers and in D2D they are backed with WIC bitmaps, which need to be format converted on every frame
- All layers are still composited on every frame, whether there is something to draw or not. This is beacuse D2D is double buffered and we need to be sure both buffers are updated. We need to provide dirty rects when flipping the swap chain to fix this.

Fixes #201
Fixes #646

grokys committed:
Initial implementation of low-level scene graph.

Whole tree is currently being updated and rendered still though and lots doesn't even build.

Commit:
Merge branch 'master' into scenegraph

grokys committed:
Removed duplicate TestRoot.

Use the one in Avalonia.UnitTests.

grokys committed:
Merge branch 'remove-duplicate-testroot' into scenegraph

 Conflicts:
	tests/Avalonia.Visuals.UnitTests/TestRoot.cs

grokys committed:
Initial implementation of scenegraph hit testing.

Based solely on control bounds as before.

grokys committed:
Updated libs with API changes.

grokys committed:
Use scene geometry for hit-testing.

This means we now correctly hit test circles etc.

grokys committed:
Fix hit testing on tab strip items.

Now that we use the scenegraph for hit testing. This fixes the tabs in
ControlCatalog.

grokys committed:
Fix initial draw.

grokys committed:
Don't need the context.Render any more.

In hit testing unit tests.

Commit:
Clip affects rendering.

Commit:
Initial impl of incremental update to SceneGraph.

Commit:
Docs and tweaking.

Commit:
Fix GeometryNode hit test.

And add to-do for RectangleNode geometry hit-test.

grokys committed:
Reset needsRender flag after rendering.

grokys committed:
Make Shape render properly again.

grokys committed:
Fixed pointerover state.

And added a test for it.

grokys committed:
Make SceneBuilder update dirty rects.

grokys committed:
Render child visual nodes explicitly.

Render them in DeferredRenderer rather than in VisualNode.

grokys committed:
Initial implementation of dirty rect drawing.

grokys committed:
Move FPS drawing out of inner loop.

grokys committed:
Removed unnecessary field.

grokys committed:
Special case for hit testing in unit tests.

When not in a unit test, only update the scene from the render loop.

grokys committed:
Prevent reentrancy in renderer.

grokys committed:
Fixed a few things.

- We don't need ConcurrentQueue in renderer
- DirtyRects coalescing was wrong
- Display number of updated rects for debug until we have dirty rect
rendering
- Animate 2 items at once in RenderTest

grokys committed:
Fix non-compiling tests.

grokys committed:
Use null _dirty as meaning "needs first update".

grokys committed:
Don't keep rendering when there's nothing to render.

grokys committed:
Simplification.

Previous code was me trying to work something out.

grokys committed:
Use correct bounds in VisualNode.

grokys committed:
Make IGeometryImpl immutable.

As it will be shared with the UI thread and the render thread.

grokys committed:
Make SharpDX dispose of its resources.

grokys committed:
Make FormattedTextImpl immutable.

Because it needs to be shared between the UI thread and the render
thread. This also required making it non-disposable like the other
graphics primitive impls.

grokys committed:
Prevent out-of-bounds in DirtyRects.Coalesce.

grokys committed:
Fix bounds clipping.

grokys committed:
Fix RectangleNode bounds.

grokys committed:
Fix geometry measure.

grokys committed:
Remove unneeded cast.

grokys committed:
Refactored VisualNode.

It now has two collections: Children and DrawOperations.

grokys committed:
WIP: Adding render layers.

grokys committed:
Add option to draw dirty rects.

grokys committed:
More work on deferred renderer.

Getting opacity layers working.

grokys committed:
Store layers in Scene.

Layers are now calculated by SceneBuilder and stored in Scene. This
means we are no longer accessing `visual.Opacity` from the render thread
(and `VerifyAccess` calls have been placed in AvaloniaObject.GetValue
and friends to ensure this cannot happen). ControlCatalog is now
rendering mostly fine.

grokys committed:
WIP: Getting resizing working.

grokys committed:
Update dirty controls in order.

grokys committed:
Use the Direct2D 1.1 render target initialization.

To prevent D2D from auto-stretching our render target to the window size
(which causes problems now we're multi-threaded) we have to move to the
D2D1.1 render target initialization. I used the info here to work out
how to do this:

https://katyscode.wordpress.com/2013/01/23/migrating-existing-direct2d-applications-to-use-direct2d-1-1-functionality-in-windows-7/
- it's not exactly simple!

grokys committed:
Handle detached visuals in dirty list.

Commit:
Make DPI scaling work again.

grokys committed:
Initialize layer backgrounds to transparent.

grokys committed:
Resize swap chain on window resize.

Rather than recreating it.

grokys committed:
A bit of work on Cairo, Skia and Gtk backends.

They're not yet working properly but they're drawing something and not
crashing.

grokys committed:
Don't clear canvas when creating drawing context.

grokys committed:
Make render tests pass again.

grokys committed:
Removed RendererMixin.

Moved immediate rendering logic into `Renderer`.

grokys committed:
Renamed Renderer -> ImmediateRenderer

grokys committed:
Fix failing tests.

grokys committed:
Share FPS drawing between renderers.

grokys committed:
Fix immediate geometry rendering in D2D.

grokys committed:
Added skia ref to RenderTest.

grokys committed:
Don't draw FPS to overlay.

grokys committed:
Don't clear GTK window on expose.

Commit:
Added setting to save frames to disk.

For debugging purposes.

Commit:
Fix skia text rendering.

We were seeing blurry text when switching pages in the ControlCatalog
using Skia. This fixes that, though I'm not entirely sure how. Thanks to
@danwalmsley for the soltuion!

Commit:
Make Window.IsVisible track window visibility.

Commit:
Refactor render layers a bit.

We be sure of the required state at the start of rendering,

Commit:
Handle deleted layers.

Commit:
Swallow exceptions in render loop.

And remove debug output statement.

Commit:
Merge branch 'master' into scenegraph

 Conflicts:
	samples/TestApplication/App.config
	samples/XamlTestApplication/App.config
	samples/XamlTestApplicationPcl/TestScrollable.cs
	tests/Avalonia.RenderTests/Media/FormattedTextImplTests.cs

Commit:
Run render tests with deferred renderer.

Make render tests run all tests with both the deferred and immediate
renderer.

Commit:
Remove Rect.operator*

It was wrong - use Rect.TransformToAABB instead.

Commit:
Use widened bounds for GeometryNode.Bounds.

Commit:
Don't call Paint for now on Win32.

It's causing stuff to crash. We will probably want to do this later to
make resizing look better though.

Commit:
Handle null event source.

Commit:
Fix RenderLayer resize.

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Windows/Avalonia.Direct2D1/Direct2D1Platform.cs
	src/Windows/Avalonia.Direct2D1/HwndRenderTarget.cs
	src/Windows/Avalonia.Direct2D1/Media/DrawingContextImpl.cs
	src/Windows/Avalonia.Direct2D1/Media/Imaging/BitmapImpl.cs
	src/Windows/Avalonia.Direct2D1/RenderTarget.cs

grokys committed:
Remove refrences to "scene graph" in Visual.

As its now a separate thing.

kekekeks commented:
Why exactly do you need direct access to `IDrawingContextImpl`? The whole point of having `DrawingContext` is to keep it's state from being inconsistent or corrupted.

kekekeks commented on src/Avalonia.Visuals/Rendering/SceneGraph/SceneBuilder.cs:
```diff
@@ -0,0 +1,322 @@
+                    {
+                        foreach (var child in visual.VisualChildren.OrderBy(x => x, ZIndexComparer.Instance))
+                        {
+                            var childNode = scene.FindNode(child) ?? CreateNode(scene, child, node);
```
Are you creating a bitmap-backed layer/node for each visual?

kekekeks commented:
It also breaks additional matrix configurations that was used by skia here https://github.com/AvaloniaUI/Avalonia/blob/089b505e01cd9daef4fd1ccb1d0ab71658bc12f7/src/Skia/Avalonia.Skia.Desktop/RenderTarget.cs#L132

grokys commented:
Yep, good question - I'll try to explain my thinking here and then we can talk about whether we should be taking a different path. I'm going to use geometry as an example here but it also applies to things like formatted text etc.

So the new renderer needs to share objects such as geometries between the UI thread and the render thread. The easiest way to do this is by using immutable state, and we already had an (almost) immutable geometry in `IGeometryImpl` (`IGeometryImpl` wasn't quite immutable - it had a mutable `Transform` property - I made that immutable. Likewise `IFormattedText` was almost immutable but not quite, so some changes had to be made).

So now we have a mutable `Geometry` and an immutable `IGeometryImpl` we need something similar for drawing contexts, something like... `DrawingContext` and `IDrawingContextImpl`. Which we already had. So I changed `IDrawingContextImpl` to take the immutable objects - this required basically 0 work,

Now regarding `DrawingContext` preventing the state of `IDrawingContextImpl` getting corrupted - when we build the scene graph, we do so via a `DrawingContext`. The current state of the drawing context is written to the scene graph node, so it can be instantly restored. The scene graph node simply writes that state to the `IDrawingContextImpl`, so the state _should_ be exactly what was calculated by `DrawingContext`.

I say _should_ because there are a couple of things this doesn't handle so well - one as you identified was the hidden state in the drawing context for skia DPI scaling. The others that I think may cause problems are geometry clips and opacity masks (not tried to implement these yet).

I'd be interested in your input in these issues - the PR is still very much a WIP and the more input we get the better.

kekekeks commented:
Well, if you want an immutability layer, it's probably better to rename the current `DrawingContext` to something else and let it manage state stack management. And then implement new `DrawingContext` for handling mutability issues.

grokys committed:
Added ncrunch 3 solution.

grokys committed:
WIP: Working on getting VisualBrush working

Don't try to render `VisualBrush`es in TileBrushImpl - visual brushes
need first to be transformed into images and then rendered as an
`ImageBrush`. To this end, renamed `TileBrushImpl` as `ImageBrushImpl`
(in D2D backend only for now as that's what I'll get working first).
Actual visual brush implementation not yet done.

grokys committed:
WIP: Delegate back to render for VisualBrushes.

`VisualBrush`es need to get delegated back to the renderer for
rendering, so added an `IVisualBrushRenderer` interface which is
implemented by the renderers, and pass this when creating drawing
contexts. Only currently implemented/tested for D2D and
`ImmediateRenderer`.

grokys committed:
Don't throw here.

It makes Direct2D hang for some reason.

grokys committed:
Added in-tree visual brush test.

Expected output is slightly wonky - need to investigate that once
stuff's working in general.

grokys committed:
WIP: More work on VisualBrush

Laying the foundations for getting `VisualBrush` working with
`DeferredRenderer`.

grokys committed:
Updated test output.

ImageMagick couldn't tell that the deferred renderer was failing the
test previously!!?!?

grokys committed:
WIP: Draw visual brush from DeferredRenderer.

It's not drawing the correct thing though, as we need to create a new
scene for the visual brush visual.

grokys committed:
Run deferred render tests on a background thread.

Run deferred render tests on a background thread to expose any threading
errors in the deferred rendering path. And threading errors found. Also
had to decrease the timeout for D2D render tests as
`renderTarget.EndDraw` is hanging when an exception is thrown between
`renderTarget.BeginDraw` and `EndDraw`.

grokys committed:
Added IImageBrush and IVisualBrush

And make an immutable `SceneImageBrush` and `SceneVisualBrush` which
clone the mutable `ImageBrush` and `VisualBrush` for use in the scene
graph, as we shouldn't be accessing mutable state from the render
thread.

grokys committed:
Got VisualBrushes working again.

grokys committed:
Make scene nodes internal.

grokys committed:
Removed random class diagram.

grokys committed:
Refactored brushes.

`SolidColorBrush` is now mutable. There are immutable versions of
brushes in the `Avalona.Media.Immutable` namespace and the static
`Brushes.Xxx` properties return these.

grokys committed:
Make LinearGradientBrush work again.

With deferred renderer.

grokys committed:
Make RadialGradientBrush work

With immediate renderer. Also added a render test and fixed some
problems with D2D implementation.

grokys committed:
Make OpacityMask work in deferred renderer.

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Gtk/Avalonia.Gtk/WindowImplBase.cs
	src/Skia/Avalonia.Skia.Android/RenderTarget.cs
	src/Skia/Avalonia.Skia.Android/SkiaRenderView.cs
	tests/Avalonia.Input.UnitTests/InputElement_HitTesting.cs
	tests/Avalonia.LeakTests/ControlTests.cs

grokys committed:
Use yellow background in geometry clip tests.

For some reason imagemagick couldn't tell that the renders were wrong
before.

grokys committed:
Added failing geometry clip hit-test.

grokys committed:
Updated RenderTest.

It has various problems under deferred renderer.

grokys committed:
Added failing test.

RenderTransforms aren't being applied correctly after initial scene
build.

grokys committed:
Fix RenderTransforms in SceneBuilder.

grokys committed:
Make geometry clips work.

Still problems with hit-testing and geometry clips aren't propogated to
new layers.

grokys committed:
Fixed geometry clip hit testing.

Updated unit test to test for previous problem.

grokys committed:
Moved some mocks into Avalonia.UnitTests.

From Avalonia.Visuals.UnitTests as they're useful generally.

grokys committed:
Handle geometry clips on child layers.

grokys committed:
Test turning geometry clipping on and off.

Seems to work!

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Avalonia.Visuals/Media/DrawingContext.cs
	src/Skia/Avalonia.Skia.iOS/RenderTarget.cs
	src/Skia/Avalonia.Skia/FramebufferRenderTarget.cs

grokys committed:
Fixed failing tests.

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Skia/Avalonia.Skia/DrawingContextImpl.cs

grokys committed:
Make cairo render tests pass.

grokys committed:
Implement drawing opacity masked bitmap on skia.

grokys committed:
Don't clear skia canvas when creating drawing ctx.

grokys committed:
Run skia render tests under NCrunch.

grokys committed:
Make a bunch of skia render tests pass.

grokys committed:
Fix Skia stream geometry bounds calculation.

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	tests/Avalonia.RenderTests/Shapes/PathTests.cs

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Gtk/Avalonia.Gtk/TopLevelImpl.cs
	tests/Avalonia.Controls.UnitTests/TopLevelTests.cs

grokys committed:
Fix failing test.

grokys committed:
Send paint and resize notifications to renderer.

grokys committed:
Tweak DeferredRenderer locking.

grokys committed:
Set dirty rect when control removed/hidden.

And added unit tests. This bug resulted in a glitch on the Carousel page
in ControlCatalog when the images were scrolled.

grokys committed:
Added missing member to interface.

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Windows/Avalonia.Direct2D1/Media/Imaging/WicBitmapImpl.cs
	tests/Avalonia.RenderTests/Avalonia.RenderTests.projitems
	tests/Avalonia.Visuals.UnitTests/VisualTree/MockRenderInterface.cs

grokys committed:
Forgot to save before last commit.

grokys committed:
Moved ncrunch files to their own dir.

Also deleted v2 ncrunch files and update .gitignore.

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	samples/RenderTest/MainWindow.xaml.cs

grokys committed:
Added doc comments for ImmediateRenderer.

grokys committed:
Reworked immediate renderer.

It now doesn't use the render loop, but instead invalidates the window
and reacts to paint events.

grokys committed:
Make windowing subsystem select the renderer.

Direct2D selects `DeferredRenderer`, the GTK backends select
`ImmediateRenderer` for now.

grokys committed:
Allow selection of ImmediateRendering on Win32.

grokys committed:
Don't need to pass visual brush renderer when creating render target.

grokys committed:
Merge remote-tracking branch 'origin/master' into scenegraph

 Conflicts:
	src/Avalonia.Visuals/Media/GradientBrush.cs
	src/Avalonia.Visuals/Media/IGradientBrush.cs
	src/Avalonia.Visuals/Media/IRadialGradientBrush.cs
	src/Avalonia.Visuals/Media/ImageBrush.cs
	src/Avalonia.Visuals/Media/LinearGradientBrush.cs
	src/Avalonia.Visuals/Media/RadialGradientBrush.cs
	src/Avalonia.Visuals/Media/VisualBrush.cs
	src/Gtk/Avalonia.Cairo/Media/DrawingContext.cs
	src/Gtk/Avalonia.Cairo/Media/ImageBrushImpl.cs
	src/Gtk/Avalonia.Cairo/Media/LinearGradientBrushImpl.cs
	src/Gtk/Avalonia.Cairo/Media/RadialGradientBrushImpl.cs
	src/Gtk/Avalonia.Cairo/Media/TileBrushes.cs
	src/Gtk/Avalonia.Cairo/Media/VisualBrushImpl.cs
	src/Skia/Avalonia.Skia/DrawingContextImpl.cs
	src/Windows/Avalonia.Direct2D1/Media/DrawingContextImpl.cs
	src/Windows/Avalonia.Direct2D1/Media/LinearGradientBrushImpl.cs
	src/Windows/Avalonia.Direct2D1/Media/RadialGradientBrushImpl.cs
	src/Windows/Avalonia.Direct2D1/Media/TileBrushImpl.cs

grokys committed:
Removed unused code.

grokys committed:
Merge branch 'master' into scenegraph

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Avalonia.Visuals/Media/DrawingContext.cs
	src/Avalonia.Visuals/Media/IDrawingContext.cs
	src/Avalonia.Visuals/Media/IDrawingContextImpl.cs
	src/Avalonia.Visuals/Platform/IDrawingContextImpl.cs
	src/Gtk/Avalonia.Cairo/Media/DrawingContext.cs
	src/Gtk/Avalonia.Cairo/Media/FormattedTextImpl.cs
	src/Skia/Avalonia.Skia/DrawingContextImpl.cs
	src/Skia/Avalonia.Skia/FormattedTextImpl.cs
	src/Windows/Avalonia.Direct2D1/Media/DrawingContextImpl.cs
	src/Windows/Avalonia.Direct2D1/Media/FormattedTextImpl.cs
	src/Windows/Avalonia.Direct2D1/RenderTarget.cs
	src/Windows/Avalonia.Direct2D1/SwapChainRenderTarget.cs
	tests/Avalonia.Controls.UnitTests/WindowTests.cs
	tests/Avalonia.Markup.Xaml.UnitTests/Xaml/StyleTests.cs

grokys committed:
Removed unused member.

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Avalonia.Base/Threading/Dispatcher.cs

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Avalonia.Controls/Primitives/PopupRoot.cs
	src/Avalonia.Controls/TopLevel.cs
	tests/Avalonia.Controls.UnitTests/WindowTests.cs
	tests/Avalonia.UnitTests/TestServices.cs

grokys committed:
Remove .vscode dir.

grokys committed:
Merge remote-tracking branch 'origin/master' into scenegraph

grokys committed:
Remove IsVisible changes from Window/Popup.

These were from a previous attempt to do what #935 does.

grokys committed:
Use correct namespace for IDrawingContextImpl.

grokys committed:
Make IFormattedTextImpl immutable.

Previously `IFormattedTextImpl` had `SetForegroundBrush` which set
mutable state. Make `FormattedText` fully mutable (before it was kinda
mutable, kinda immutable) and create immutable `IFormattedTextImpl`s on
demand.

grokys committed:
Added some docs.

And a few missing child scene properties.

grokys committed:
FormattedText spans can be null.

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Avalonia.Visuals/Media/FormattedTextStyleSpan.cs
	src/Avalonia.Visuals/Media/Typeface.cs
	src/Avalonia.Visuals/Rendering/RendererMixin.cs
	src/Gtk/Avalonia.Cairo/Media/StreamGeometryImpl.cs
	src/Windows/Avalonia.Direct2D1/Direct2D1Platform.cs
	tests/Avalonia.Input.UnitTests/InputElement_HitTesting.cs
	tests/Avalonia.Layout.UnitTests/FullLayoutTests.cs
	tests/Avalonia.UnitTests/TestServices.cs
	tests/Avalonia.Visuals.UnitTests/Media/TypefaceTests.cs

grokys committed:
Merge branch 'port-immediaterenderer' into scenegraph

 Conflicts:
	src/Avalonia.Visuals/Platform/IRenderTarget.cs
	src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs
	src/Gtk/Avalonia.Cairo/Media/DrawingContext.cs
	src/Gtk/Avalonia.Cairo/Media/ImageBrushImpl.cs
	src/Shared/RenderHelpers/TileBrushImplHelper.cs
	src/Skia/Avalonia.Skia/BitmapImpl.cs
	src/Skia/Avalonia.Skia/DrawingContextImpl.cs
	src/Windows/Avalonia.Direct2D1/Direct2D1Platform.cs
	src/Windows/Avalonia.Direct2D1/Media/Imaging/RenderTargetBitmapImpl.cs
	src/Windows/Avalonia.Direct2D1/RenderTarget.cs
	src/Windows/Avalonia.Direct2D1/SwapChainRenderTarget.cs
	src/Windows/Avalonia.Win32/Win32Platform.cs
	tests/Avalonia.UnitTests/TestServices.cs
	tests/Avalonia.Visuals.UnitTests/VisualTests.cs
	tests/Avalonia.Visuals.UnitTests/VisualTree/VisualExtensionsTests_GetVisualsAt.cs

grokys committed:
Fix merge errors.

grokys committed:
Moved hit testing tests.

grokys committed:
Added ImmediateRenderer hit tests.

grokys committed:
Merge branch 'fixes/fix-getvisualroot' into scenegraph

 Conflicts:
	src/Skia/Avalonia.Skia.iOS/RenderTarget.cs
	src/Skia/Avalonia.Skia.iOS/WindowDrawingContextImpl.cs
	src/Skia/Avalonia.Skia/BitmapImpl.cs
	src/Windows/Avalonia.Direct2D1/Direct2D1Platform.cs
	src/Windows/Avalonia.Direct2D1/Media/Imaging/WicBitmapImpl.cs
	tests/Avalonia.UnitTests/TestRoot.cs
	tests/Avalonia.Visuals.UnitTests/Rendering/DeferredRendererTests_HitTesting.cs
	tests/Avalonia.Visuals.UnitTests/Rendering/ImmediateRendererTests_HitTesting.cs
	tests/Avalonia.Visuals.UnitTests/VisualTree/VisualExtensionsTests_GetVisualsAt.cs

grokys committed:
Removed some small diffs from master.

grokys committed:
Added RenderScaling doc comment.

grokys committed:
Always allow D2D for the moment.

grokys committed:
This needs to be settable.

grokys committed:
Made some classes internal and added docs.

grokys committed:
Merge branch 'master' into scenegraph

 Conflicts:
	src/Shared/SharedAssemblyInfo.cs

grokys committed:
Merge remote-tracking branch 'origin/master' into scenegraph

grokys committed:
Expose Visuals internals to render tests.

grokys committed:
Added failing custom renderer test.

Deferred rendering does not yet support opacity from `Visual.Render`.

grokys committed:
Implemented deferred opacity drawing.

grokys committed:
Added failing render test.

Deferred rendering does not yet support clipping in `Visual.Render`.

grokys committed:
Adjusted expected test output.

The output images were off-center.

grokys committed:
Added failing custom render tests.

Deferred rendering does not yet support geometry clipping or opacity masks from `Visual.Render`.

grokys committed:
Implemented more deferred rendering.

Implemted geometry clips and opacity masks.

grokys committed:
Removed duplicate using statement.

grokys committed:
Merge branch 'master' into scenegraph

grokys committed:
Fixed expected skia output.

grokys committed:
Fix possible null reference exceptions.

grokys committed:
Removed files deleted on master.

grokys committed:
Port Direct2D code from master.

grokys committed:
Fix DPI scaling with Skia.

grokys committed:
Merge branch 'fixes/fix-interop-samples' into scenegraph

grokys committed:
Merge branch 'fixes/fix-interop-samples' into scenegraph

grokys committed:
Make D3D interop sample work.

By turning off deferred rendering.

Commit:
Merge branch 'master' into scenegraph

danwalmsley commented:
@grokys did you find a way to manually test this build with other apps like Core2d or AvalonStudio, I failed miserably trying to replace the files in the nuget directory.

grokys committed:
Removed bad project references.

grokys committed:
Don't remove root layer when hiding root.

grokys committed:
Start/stop the renderer on window show/hide.

grokys committed:
Merge remote-tracking branch 'origin/master' into scenegraph

 Conflicts:
	src/Skia/Avalonia.Skia/FramebufferRenderTarget.cs
	src/Windows/Avalonia.Direct2D1/Direct2D1Platform.cs

grokys committed:
Set background on MenuItem and TreeViewItem

Now we have mouse transparency with the deferred renderer, we need to make sure there's a transparent background to make them hittable.

grokys committed:
Make BoundsTracker work with deferred renderer.

This needs to be redone: see #1026.

grokys committed:
Set background for CheckBox and RadioButton

Now we have mouse transparency with the deferred renderer, we need to make sure there's a transparent background to make them hittable.

grokys committed:
GeometryNode.Pen may be null.

grokys committed:
Check for null from Scene.FindNode.

Scene.FindNode(layerRoot) may return null if the layer root has been hidden.

grokys committed:
Convert brushes to immutable brushes in FormattedTextImpl.

They need to be accessed from the render thread.

danwalmsley commented:
just a reminder of currently outstanding issues on this branch that I am aware of:
- [ ] When window is minimized constantly throwing an exception because WindowFrameBuffer constructor is called with width / height = 0.

grokys committed:
Change cast to `as` operator.

Damn brane.

Commit:
Merge branch 'master' into scenegraph

grokys committed:
Don't try to create a framebuffer of size 0.

grokys commented on src/Windows/Avalonia.Win32/FramebufferManager.cs:
```diff
@@ -25,7 +25,7 @@ public ILockedFramebuffer Lock()
             var width = rc.right - rc.left;
             var height = rc.bottom - rc.top;
-            if (_fb == null || _fb.Width != width || _fb.Height != height)
+            if ((_fb == null || _fb.Width != width || _fb.Height != height) && width > 0 && height > 0)
```
@kekekeks this is an attempt to fix the problem that AvalonaStudio was seeing when the window was maximized and `width`/`height` of 0 was passed to `WindowFramebuffer`. I'm not sure if it's the correct fix.

kekekeks replied:
By doing this  you are canceling framebuffer resize. I think that you should cancel render pass completely, since window is not visible anyway.

grokys committed:
Fix scrollwheel scrolling in TreeView/ListBox.

Commit:
Merge branch 'master' into scenegraph

danwalmsley commented:
@grokys since you fixed the Fix scrollwheel on Treeview / Listbox...
Listbox seems to have a different background before, see the Recent solutions list in avalon studio, it has a lighter background which it didnt before.

I've noticed it in other places too.

![image](https://user-images.githubusercontent.com/4672627/27955777-564fb68a-630e-11e7-81c1-f0d6d0b1ea41.png)

Commit:
Merge branch 'master' into scenegraph

kekekeks commented on src/Skia/Avalonia.Skia.Desktop/RenderTarget.cs:
```diff
@@ -127,9 +126,8 @@ public override DrawingContext CreateDrawingContext()
-                new DrawingContext(
-                    new WindowDrawingContextImpl(this), Matrix.CreateScale(scale, scale));
+            // TODO: Broken DPI scaling on skia here. Fix this.
+            var result = new WindowDrawingContextImpl(this);
```
I don't think that WindowRenderTarget (or RenderTarget in general) is used anymore. We should remove this file.

kekekeks commented on src/Skia/Avalonia.Skia/DrawingContextImpl.cs:
```diff
@@ -20,12 +22,16 @@ internal class DrawingContextImpl : IDrawingContextImpl
 
         public DrawingContextImpl(
             SKCanvas canvas,
+            double dpiX,
```
If I remember correctly, you've wanted DPI to be stored as `Vector`

grokys replied:
Yep, this one slipped through the cracks.

kekekeks commented on src/Skia/Avalonia.Skia/FormattedTextImpl.cs:
```diff
@@ -616,6 +616,7 @@ private void SetForegroundBrush(IBrush brush, int startIndex, int length)
 
             if (brush != null)
             {
+                brush = (brush as IMutableBrush)?.ToImmutable() ?? brush;
```
See? It's better to move this check to `ToImmutable`.

kekekeks commented on src/Windows/Avalonia.Win32/Win32Platform.cs:
```diff
@@ -200,7 +211,9 @@ public IPopupImpl CreatePopup()
 
         public IRenderer CreateRenderer(IRenderRoot root, IRenderLoop renderLoop)
```
This should be moved to `ITopLevelImpl` since we can mix multiple toplevel implementations in the same app (e. g. WPF embedded avalonia control showing avalonia-backed `Popup`)

kekekeks commented on tests/Avalonia.Input.UnitTests/MouseDeviceTests.cs:
```diff
@@ -0,0 +1,87 @@
+            using (TestApplication(renderer.Object))
+            {
+                var inputManager = InputManager.Instance;
+                var mouseDevice = AvaloniaLocator.Current.GetService<IMouseDevice>();
```
`MouseDevice` was moved to `IInputRoot`, if I remember correctly

kekekeks reviewed (changes requested):
Marked issues that are outside of the scope of this PR with `[FUTURE]` prefix. Please, add `TODO:` in code or something.

Commit:
Merge branch 'master' into scenegraph

danwalmsley commented:
This branch may also have the following issues:

- [ ] Mouse transparency preventing context menu closing (command is initiated)
- [ ] Mouse transparency preventing pointerover on drop down items, unless right over the textblock part.

grokys committed:
Newly merged tests must be async.

grokys committed:
Give background to DropDownItem and ListBoxItem.

grokys committed:
Fix ScrollViewer background.

It was being set twice.

grokys committed:
Fixed ScrollViewer background correctly.

Removed it from the wrong place in the previous commit.

grokys commented on src/Avalonia.Visuals/Rendering/SceneGraph/Scene.cs:
```diff
@@ -0,0 +1,184 @@
+        /// <summary>
+        /// Gets an ID identifying the scene. This is incremented each time the scene is cloned.
+        /// </summary>
+        public int Id { get; }
```
True, perhaps `Generation` would be a better name.

grokys committed:
Merge remote-tracking branch 'origin/master' into scenegraph

grokys commented on src/Avalonia.Controls/Window.cs:
```diff
@@ -225,8 +225,14 @@ public void Close(object dialogResult)
+
             using (BeginAutoSizing())
             {
+                Renderer?.Stop();
```
Do you mean that we _also_ need to react to `WindowState` changes, or that some other form of window state should _replace_ `Hide`?

grokys commented on src/Avalonia.Visuals/Rendering/SceneGraph/BrushDrawOperation.cs:
```diff
@@ -0,0 +1,62 @@
+        /// </summary>
+        /// <param name="brush">The brush.</param>
+        /// <returns>An immutable brush</returns>
+        protected IBrush ToImmutable(IBrush brush)
```
The `ToImmutable` method that it calls isn't an extension method - it's defined by `IMutableBrush` and is a method on the brush itself.

kekekeks replied:
Would be nice to have something like `EnsureImmutable` then

grokys commented on src/Avalonia.Visuals/Rendering/SceneGraph/DeferredDrawingContextImpl.cs:
```diff
@@ -0,0 +1,391 @@
+        /// <inheritdoc/>
+        public void Clear(Color color)
+        {
+            // Cannot clear a deferred scene.
```
Yeah we could add a separate clear operation, but it's unclear how it would work with layers... Something we should discuss apart from the PR if it's necessary IMO.

But yeah, I think a more sensible move for now is to remove `Clear` from `IDrawingContext`. WPF for example doesn't have it there.

grokys replied:
We actually need `Clear` on `IDrawingContextImpl` for the moment (note that it's not exposed to controls in `DrawingContext`) as it's needed to clear a the dirty rect of a layer before rendering.

danwalmsley commented:
![image](https://user-images.githubusercontent.com/4672627/28410616-2068f342-6d36-11e7-9548-a696b0ccbcd6.png)
Just got this exception on scenegraph :(

will keep an eye out for it.

Commit:
Merge branch 'master' into scenegraph

Commit:
Merge branch 'master' into scenegraph

grokys committed:
Fix merge error.

grokys committed:
MouseDevice now exposed on TopLevel.

grokys committed:
Added TODOs.

grokys committed:
Removed unused file.

grokys committed:
Use Vector for DPI.

Still needs changing in `IPlatformRenderInterface` but I'd prefer to make that change in a separate PR.

grokys committed:
Add Brush/Pen.ToImmutable() extension method.

grokys committed:
Scene.Id -> Scene.Generation.

grokys commented on src/Avalonia.Visuals/Rendering/SceneGraph/SceneBuilder.cs:
```diff
@@ -0,0 +1,389 @@
+                using (context.PushPostTransform(m))
+                using (context.PushTransformContainer())
+                {
+                    var startLayer = opacity < 1 || visual.OpacityMask != null;
```
Unfortunately doing this at the moment causes RenderTest to run at 1 FPS on D2D. I think we should do this after we fix the use of WIC bitmaps for layers in D2D.

grokys committed:
Throw NotSupportedException.

And add a description of why it's not supported.

grokys committed:
Removed IRendererFactory.

Make the `ITopLevelImpl` create the renderer instead.

grokys committed:
Merge remote-tracking branch 'origin/master' into scenegraph

grokys committed:
Fix merge error.

grokys committed:
Remove IRenderFactory from mobile platforms.

grokys commented:
@kekekeks how's this looking to you now?

grokys committed:
Merge branch 'master' into scenegraph
